core/src/main/java/org/elasticsearch/action/index/IndexRequest.java

```diff
@@ -467,7 +469,19 @@ public IndexRequest version(long version) {
 
     @Override
     public long version() {
-        return this.version;
+        return resolveVersionDefaults();
```

Comment:
nit: this is tricky as one can do `setVersion(MATCH_ANY)` and get `MATCH_DELETED` back. I think it's fine but let's just add a note about it?